app/src/org/commcare/sync/FirebaseMessagingDataSyncer.java

```diff
@@ -0,0 +1,334 @@
+                    recreateActivity = true;
+                    break;
+                case MENU:
+                    recreateActivity = true;
+                    break;
+            }
+            commcareSession.syncState();
+            if (recreateActivity) {
+                activity.recreate();
+            }
+        }
+        catch(RuntimeException e){
+            activity.setResult(RESULT_CANCELED);
+            activity.finish();
+        }
+    }
+
+    private boolean checkUserAndDomain(User user, String payloadUsername, String payloadDomain) {
+        if(payloadUsername != null && payloadDomain != null){
+            String loggedInUsername = user.getUsername();
+            String userDomain = removeServerUrlFromUserDomain(HiddenPreferences.getUserDomain());
+            return payloadUsername.equalsIgnoreCase(loggedInUsername) && payloadDomain.equalsIgnoreCase(userDomain);
+        }
+        return false;
+    }
+
+    private AppNavigationStates getAppNavigationState() {
+        if (CommCareActivity.currentActivity instanceof FormEntryActivity)
+            return FORM_ENTRY;
+        else if (CommCareActivity.currentActivity instanceof EntitySelectActivity)
+            return ENTITY_SELECTION;
+        else if (CommCareActivity.currentActivity instanceof EntityDetailActivity)
+            return ENTITY_DETAIL;
+        else if (CommCareActivity.currentActivity instanceof StandardHomeActivity)
+            return HOME_SCREEN;
+        else if (CommCareActivity.currentActivity instanceof MenuActivity)
+            return MENU;
+        return OTHER;
+    }
+
+    // This method is responsible for informing the User and scheduling a sync for when it's safe
+    private void informUserAboutPendingSync(CommCareActivity activity, FCMMessageData fcmMessageData) {
+        activity.runOnUiThread(() ->
+                activity.alertPendingSync(fcmMessageData)
+        );
+    }
+
+    @Override
+    public void connectTask(CommCareTask task) {
+        this.currentTask = task;
+    }
+
+    @Override
+    public void startBlockingForTask(int id) {
+        // In case a background sync is already ongoing, we ignore the new request silently
+        if (CommCareSessionService.sessionAliveLock.isLocked()) {
+            currentTask.cancel(true);
+        }
+        mPinnedNotificationProgress = new PinnedNotificationWithProgress(context,
+                "sync.communicating.title","sync.progress.starting", -1);
+
+        // Disable any pending sync
+        disablePendingSyncs();
+    }
+
+    private void disablePendingSyncs() {
+        HiddenPreferences.setPendingSyncRequestFromServer(false);
+        HiddenPreferences.setPostFormSubmissionSyncNeeded(false);
+    }
+
+    @Override
+    public void stopBlockingForTask(int id) {
+        // Acquire lock to finalize sync. This is not expected to fail, but in case it happens
+        // we will continue
+        CommCareSessionService.sessionAliveLock.tryLock();
```

Comment:
this gets called even when sync is cancelled or finished in which case we don't want to get a lock. We should be applying the lock in `onPreExecute` and releasing it on `OnCancelled` and in `onPostExecute`

Reply:
this lock is not for the sync itself but for post sync tasks, basically to shield [`processReturnFromBackgroundSync`](https://github.com/dimagi/commcare-android/blob/1194304aa837e2e412591461fa7adb3207797a55/app/src/org/commcare/sync/FirebaseMessagingDataSyncer.java#L169C41-L169C41) from new sync requests.